creme/neighbors/base_neighbors.py

```diff
@@ -0,0 +1,208 @@
+        self._X = None
+        self._y = None
+        self._is_initialized = False
+
+        return self
+
+    def append(self, x: np.ndarray, y: base.typing.Target) -> 'KNeighborsBuffer':
+        """Add a (single) sample to the sample window.
+
+        x : 1D-array of feature for a single sample.
+
+        y : The target data for a single sample.
+
+        Raises
+        ------
+        ValueError: If at any moment, a sample with a different number of
+        attributes than that of the n_attributes parameter is passed, a
+        ValueError is raised.
+
+        TypeError: If the buffer type is altered by the user, or is not
+        correctly initialized, a TypeError may be raised.
+
+        """
+        if not self._is_initialized:
+            self._n_features = get_dimensions(x)[1]
+            self._n_targets = get_dimensions(y)[1]
+            self._configure()
+
+        if self._n_features != get_dimensions(x)[1]:
+            raise ValueError("Inconsistent number of features in X: {}, previously observed {}.".
+                             format(get_dimensions(x)[1], self._n_features))
+
+        self._X[self._next_insert, :] = x
+        self._y[self._next_insert] = y
+
+        slot_replaced = self._imask[self._next_insert]
+
+        # Update the instance storing logic
+        self._imask[self._next_insert] = True  # Mark slot as filled
+        self._next_insert = self._next_insert + 1 if self._next_insert < self.window_size - 1 \
+            else 0
+
+        if slot_replaced:  # The oldest sample was replaced (complete cycle in the buffer)
+            self._oldest = self._next_insert
+        else:  # Actual buffer increased
+            self._size += 1
+
+        return self
+
+    def pop(self) -> typing.Union[typing.Tuple[np.ndarray, base.typing.Target], None]:
+        """Remove and return the most recent element added to the buffer. """
+        if self.size > 0:
+            self._next_insert = self._next_insert - 1 if self._next_insert > 0 else \
+                self.window_size - 1
+            x, y = self._X[self._next_insert], self._y[self._next_insert]
+            self._imask[self._next_insert] = False  # Mark slot as free
+            self._size -= 1
+
+            return x, y
+        else:
+            return None
+
+    def popleft(self) -> typing.Union[typing.Tuple[np.ndarray, base.typing.Target], None]:
+        """ Remove and return the oldest element in the buffer. """
+        if self.size > 0:
+            x, y = self._X[self._oldest], self._y[self._oldest]
+            self._imask[self._oldest] = False  # Mark slot as free
+            self._oldest = self._oldest + 1 if self._oldest < self.window_size - 1 else 0
+            if self._oldest == self._next_insert:
+                # Shift circular buffer and make its starting point be the index 0
+                self._oldest = self._next_insert = 0
+            self._size -= 1
+
+            return x, y
+        else:
```

Comment:
No need for the `else`.

Reply:
Changed in commit [64f91a8](https://github.com/creme-ml/creme/pull/323/commits/64f91a803c1bd8871fdb4e01694f6d1be14d8ad8).